base/ltclass.dtx

```diff
@@ -1102,6 +1103,99 @@
 %
 %
 %
+%  \begin{macro}{
+%    \IfPackageLoadedT,\IfPackageLoadedF,
+%    \IfPackageAtLeastT,\IfPackageAtLeastF,
+%    \IfClassAtLeastT,\IfClassAtLeastF,
+%    \IfFileAtLeastT,\IfFileAtLeastF,
+%    \IfFormatAtLeastT,\IfFormatAtLeastF,
+%    \IfPackageLoadedWithOptionsT,\IfPackageLoadedWithOptionsF,
+%    \IfClassLoadedT,\IfClassLoadedF,
+%    \IfClassLoadedWithOptionsF,\IfClassLoadedWithOptionTF
+%   }
+%    A few more conditionals for convenience
+% \changes{v1.5k}{2024/04/10}{Provide T and F conditionals not just TF
+%    (gh/1262)}
+%    \begin{macrocode}
+%</2ekernel>
+%<*2ekernel|latexrelease>
+%<latexrelease>\IncludeInRelease{2024/06/01}%
+%<latexrelease>                 {\IfPackageLoadedT}{More conditionals}%
+\def\IfPackageLoadedT #1#2{\IfPackageLoadedTF{#1}{#2}{}}
+\def\IfPackageLoadedF   #1{\IfPackageLoadedTF{#1}{}}
+\def\IfClassLoadedT   #1#2{\IfClassLoadedTF{#1}{#2}{}}
+\def\IfClassLoadedF     #1{\IfClassLoadedTF{#1}{}}
+\def\IfPackageAtLeastT#1#2#3{\IfPackageAtLeastTF{#1}{#2}{#3}{}}
+\def\IfPackageAtLeastF  #1#2{\IfPackageAtLeastTF{#1}{#2}{}}
+\def\IfClassAtLeastT  #1#2#3{\IfClassAtLeastTF{#1}{#2}{#3}{}}
+\def\IfClassAtLeastF    #1#2{\IfClassAtLeastTF{#1}{#2}{}}
+\def\IfFileAtLeastT   #1#2#3{\IfFileAtLeastTF{#1}{#2}{#3}{}}
+\def\IfFileAtLeastF     #1#2{\IfFileAtLeastTF{#1}{#2}{}}
+\def\IfFormatAtLeastT   #1#2{\IfFormatAtLeastTF{#1}{#2}{}}
+\def\IfFormatAtLeastF     #1{\IfFormatAtLeastTF{#1}{}}
+\def\IfPackageLoadedWithOptionsT #1#2#3{\IfPackageLoadedWithOptionsTF{#1}{#2}{#3}{}}
+\def\IfPackageLoadedWithOptionsF   #1#2{\IfPackageLoadedWithOptionsTF{#1}{#2}{}}
+\def\IfClassLoadedWithOptionsT #1#2#3{\IfClassLoadedWithOptionsTF{#1}{#2}{#3}{}}
+\def\IfClassLoadedWithOptionsF   #1#2{\IfClassLoadedWithOptionsTF{#1}{#2}{}}
+%    \end{macrocode}
+%    
+%  \begin{macro}{\IfFileLoadedTF,\IfFileLoadedT,\IfFileLoadedF}
+%    These three commands haven't been there at all in the past.
+% \changes{v1.5k}{2024/04/10}{Provide \cs{IfFileLoadedTF} and variants
+%    (gh/1222)}
+%    \begin{macrocode}
+\def\IfFileLoadedTF#1{%
+  \expandafter\ifx\csname ver@#1\endcsname\relax
+    \expandafter\@secondoftwo
+  \else
+    \expandafter\@firstoftwo
+  \fi}
+\def\IfFileLoadedT#1#2{\IfFileLoadedTF{#1}{#2}{}}
+\def\IfFileLoadedF  #1{\IfFileLoadedTF{#1}{}}
+%    \end{macrocode}
+%    For rollback/rollforward pretend everything was available since
+%    the beginning of dawn.
+%    \begin{macrocode}
+%</2ekernel|latexrelease>
+%<latexrelease>\EndIncludeInRelease
+%<latexrelease>\IncludeInRelease{0000/00/00}%
+%<latexrelease>                 {\IfPackageLoadedT}{More conditionals}%
+%<latexrelease>
+%<latexrelease>\def\IfPackageLoadedT #1#2{\IfPackageLoadedTF{#1}{#2}{}}
```

Comment:
shouldn't this be undefining all these?

Reply:
It is what we normally do in cases like this. We keep these kind of definitions around because in case of a rollback  we want them still available if used in packages that in themselves do not rollback, e.g. you have package abc that was available in 2010 but now you get the 2024 version that has \IfPackageLoadedT and in case of a rollback to 2015 you still run the 2024 package version, so it better keeps that command around in the hope that it does still work.

what additional text got lost or where?

Reply:
> it is what we normally do in cases like this.
really? I'm not aware of anywhere else where we "rollback" to identical code to the new version. But I didn't search for any:-) Mostly in 0000/00/00 blocks I see things like `\let\obeyedline\@undefined`

the text of my comment originally got posted with just one line, despite the fact I'd written something like the text you see now. As I say I don't mind, but that argument could apply to any rollback it may break packages using newer stuff but there's no clear right and wrong here and I guess undefining won't actually help any use cases so I'm happy to go with it as it is.

Reply:
> really? I'm not aware of anywhere else where we "rollback" to identical code to the new version.

For example, the rollback code for

- [`\If(Class|Package)Loaded(WithOptions)TF`](https://github.com/latex3/latex2e/blob/96fcbe83fcec7b3bcd720a0c70b1db7dd9607c62/base/ltclass.dtx#L1072-L1100)
- [`\IfExplAtLeastTF`](https://github.com/latex3/latex2e/blob/96fcbe83fcec7b3bcd720a0c70b1db7dd9607c62/base/ltexpl.dtx#L522-L540)

Reply:
@muzimuzhi  ah thanks:-)